README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->
![Example](./doc/pagedGrid.gif)

This is an animated grid view with pagination.  Implemented by SliverGrid.
If you need an animated list view please check in flutter.  
Or if you need an animated list view with pagination you can easily give your grid item a large with to make it a list.

Flutter doesn't have an animated GridView, when you add or remove an element from a grid,
it is hard to know where you added or removed.

## Features

1. Animated sliver Grid view.
2. Pagination, scroll down to fetch new page.
3. Easy to pop a widget from inside a grid or return it back.

## Usage

```dart
    GridController gridController = GridController(pageFetcher: fetchList);
    ...
    PagedGrid(
        maxCrossAxisExtent: 200,
        childAspectRatio: 1.8,
        gridController: gridController,
        itemBuilder: (ctxt, item, index) {
          Widget row = itemView(index: index, data: item);
          return row;
        });
```

```dart
    TextDivider(child: const Text('Divider'))
```


CHANGELOG.md
## 0.5.0

* Animated paged grid view.

## 0.5.2

* TextDivider


test/marine_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:marine/marine.dart';

void main() {
  test('adds one to input values', () {
    final pagedGrid = PagedGrid(
      gridController: GridController(
        pageFetcher: () async {
          return [];
        },
      ),
      itemBuilder: (BuildContext context, item, int index) {
        return Container();
      },
      maxCrossAxisExtent: 200,
    );
  });
}


lib/pagedgrid/pagedGrid.dart
import 'package:flutter/material.dart';

import 'gridController.dart';

class PagedGrid<T> extends StatelessWidget {
  const PagedGrid(
      {Key? key,
      required this.itemBuilder,
      required this.gridController,
      required this.maxCrossAxisExtent,
      this.childAspectRatio = 1})
      : super(key: key);

  final GridController gridController;
  final Widget Function(BuildContext context, T item, int index) itemBuilder;
  final double maxCrossAxisExtent;
  final double childAspectRatio;

  @override
  Widget build(BuildContext context) {
    return ValueListenableBuilder<List>(
      valueListenable: gridController.items,
      builder: (context, value, child) {
        return SliverGrid.extent(
          maxCrossAxisExtent: maxCrossAxisExtent,
          childAspectRatio: childAspectRatio,
          children: [
            for (int i = 0; i < value.length; i++)
              PagedGridItem(
                heroTag: i + 1,
                gridController: gridController,
                lastBack:
                    i == value.length - 1 ? gridController.fetchPage : null,
                itemBuilder: () {
                  return itemBuilder(context, value[i], i);
                },
              ),
            ValueListenableBuilder(
                valueListenable: gridController.loading,
                builder: (context, value, child) {
                  if (gridController.loading.value) {
                    return Center(
                      child: Builder(builder: (context) {
                        return const CircularProgressIndicator.adaptive();
                      }),
                    );
                  } else if (gridController.items.value.isEmpty) {
                    return const Center(
                      child: Text("No item found."),
                    );
                  }
                  return Container();
                }),
          ],
        );
      },
    );
  }
}

class PagedGridItem extends StatelessWidget {
  final Object? heroTag;
  final Widget? child;
  final Widget Function() itemBuilder;
  final Function()? lastBack;
  final GridController gridController;
  const PagedGridItem({
    Key? key,
    this.heroTag,
    this.child,
    required this.itemBuilder,
    this.lastBack,
    required this.gridController,
  }) : super(key: key);
  @override
  Widget build(BuildContext context) {
    if (lastBack != null) {
      WidgetsBinding.instance?.addPostFrameCallback((_) {
        lastBack!();
      });
    }

    return Hero(
        tag: heroTag ?? UniqueKey(),
        child: ValueListenableBuilder(
            valueListenable: gridController.animatedIndex,
            builder: (context, value, child) {
              bool isAnimated = gridController.animatedIndex.value == heroTag;
              bool isReverseAnimated =
                  gridController.animatedIndex.value == -(heroTag as num);
              Widget kid = itemBuilder();
              if (isAnimated) {
                return animatedItem(kid: kid, reverse: false);
              } else if (isReverseAnimated) {
                return animatedItem(kid: kid, reverse: true);
              }
              return kid;
            }));
  }

  Widget animatedItem({required Widget kid, bool reverse = false}) {
    return TweenAnimationBuilder<double>(
      child: kid,
      duration: gridController.duration!,
      tween: Tween<double>(begin: 0, end: 1),
      builder: (context, value, child) {
        return Transform.scale(
          scale: reverse ? (1 - value) : value,
          child: kid,
        );
      },
    );
  }
}


lib/pagedgrid/gridController.dart
import 'package:flutter/material.dart';

typedef PageBuilder = List<Widget> Function(int length);

class GridController<T> {
  GridController({required this.pageFetcher, this.duration}) {
    if (duration == null) this.duration = Duration(seconds: 1);
  }
  final ValueNotifier<bool> loading = ValueNotifier(false);
  final ValueNotifier<List> items = ValueNotifier<List>([]);
  final ValueNotifier<int> animatedIndex = ValueNotifier(0);
  final Future<List> Function() pageFetcher;
  Duration? duration;

  dispose() {
    loading.dispose();
    items.dispose();
    animatedIndex.dispose();
  }

  refresh() {
    items.value.clear();
    fetchPage();
    items.notifyListeners();
  }

  addItems(List itemList) {
    items.value = items.value + itemList;
  }

  addItem(item) {
    animatedIndex.value = 1;
    items.value = [item] + items.value;

    Future.delayed(duration!, () {
      animatedIndex.value = 0;
    });
  }

  fetchPage() async {
    loading.value = true;
    // await Future.delayed(Duration(seconds: 1));
    List page = await pageFetcher();
    if (page.isNotEmpty) addItems(page);
    loading.value = false;
  }

  tagOfItem({required T data, bool Function(T t1, T t2)? comparator}) {
    int tag = -1;
    if (comparator == null) {
      if (items.value.any((element) => element == data)) {
        tag = items.value.indexOf(data);
        tag += 1;
      }
    } else {
      if (items.value.any((element) => comparator(element, data))) {
        tag = items.value.indexWhere((element) => comparator(element, data));
        tag += 1;
      }
    }

    return tag;
  }

  showItem(
      {required Widget item,
      required BuildContext context,
      required T data,
      bool Function(T t1, T t2)? comparator}) {
    int tag = tagOfItem(data: data, comparator: comparator);

    Navigator.of(context).push(new PageRouteBuilder(
        opaque: false,
        barrierDismissible: true,
        barrierColor: Colors.black26.withOpacity(0.6),
        pageBuilder: (BuildContext context, _, __) {
          return Container(
              alignment: Alignment.center,
              child: Hero(tag: tag, child: FittedBox(child: item)));
        }));
  }

  deleteItem({required T data, bool Function(T t1, T t2)? comparator}) {
    int tag = tagOfItem(data: data, comparator: comparator);
    if (tag < 1) return;
    animatedIndex.value = -tag;
    Future.delayed(duration!, () {
      animatedIndex.value = 0;
      items.value.removeAt(tag - 1);
      items.notifyListeners();
      print(items.value[0]);
    });
  }
}


example/lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:marine/marine.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Paged Grid',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Paged Grid Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  ScrollController scrollController = ScrollController();
  late GridController gridController;

  @override
  void initState() {
    super.initState();
    gridController = GridController(pageFetcher: fetchList);
    gridController.fetchPage();
  }

  @override
  void dispose() {
    super.dispose();
    gridController.dispose();
    scrollController.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: CustomScrollView(
          controller: scrollController,
          slivers: [
            const SliverToBoxAdapter(
              child:
                  SizedBox(height: 150, child: Text("Paged grid view Example")),
            ),
            SliverToBoxAdapter(
              child: TextDivider(child: const Text('Divider')),
            ),
            pagedGrid,
          ],
        ),
      ),
      floatingActionButton: floatingButton,
    );
  }

  Widget get pagedGrid {
    return PagedGrid(
        maxCrossAxisExtent: 200,
        childAspectRatio: 1.8,
        gridController: gridController,
        itemBuilder: (ctxt, item, index) {
          Widget row = itemView(index: index, data: item);
          return row;
        });
  }

  Widget itemView(
      {int index = 0, bool standlone = false, bool isAdd: false, data}) {
    Widget result = Container(
      alignment: Alignment.center,
      width: 200,
      height: 200,
      child: Card(
        child: Column(
          children: [
            Row(
              children: [
                const Text("index"),
                const SizedBox(width: 10),
                Text("$index")
              ],
            ),
            Row(
              children: [
                const Text("data"),
                const SizedBox(width: 10),
                Text("${data['data'] ?? 'unk'}")
              ],
            ),
            if (!standlone)
              TextButton.icon(
                  onPressed: () {
                    gridController.showItem(
                        data: data,
                        item:
                            itemView(index: index, standlone: true, data: data),
                        context: context);
                  },
                  icon: const Icon(Icons.more_horiz),
                  label: const Text("DETAIL")),
            if (standlone)
              Row(
                children: [
                  TextButton(
                      onPressed: () {
                        Navigator.pop(context);
                      },
                      child: const Text("BACK")),
                  TextButton(
                      onPressed: () {
                        Navigator.pop(context);
                        gridController.deleteItem(data: data);
                      },
                      child: const Text("DELETE")),
                ],
              ),
            if (isAdd)
              TextButton(
                  onPressed: () {
                    gridController.addItem({"data": "new"});
                    Navigator.pop(context);
                  },
                  child: const Text("ADD")),
          ],
        ),
      ),
    );
    return result;
  }

  Future<List> fetchList() async {
    if (gridController.items.value.length > 100) return [];
    var ran = Random();
    List items = [];
    for (int i = 0; i < 2; i++) {
      Map item = {"data": ran.nextInt(10000)};
      items.add(item);
    }
    await Future.delayed(const Duration(milliseconds: 500));
    return items;
  }

  Widget get floatingButton {
    return Align(
      alignment: const Alignment(1, 1),
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          SizedBox(
            height: 40,
            width: 40,
            child: FloatingActionButton(
              heroTag: -1,
              onPressed: () {
                gridController.showItem(
                    data: {},
                    item: itemView(
                        index: -1, standlone: true, isAdd: true, data: {}),
                    context: context);
              },
              child: const Icon(Icons.add),
            ),
          ),
          const SizedBox(
            height: 10,
          ),
          SizedBox(
            height: 40,
            width: 40,
            child: FloatingActionButton(
              heroTag: null,
              onPressed: () {
                scrollController.animateTo(0,
                    duration: const Duration(milliseconds: 500),
                    curve: Curves.easeOut);
              },
              child: const Icon(Icons.arrow_upward),
            ),
          ),
          const SizedBox(
            height: 10,
          ),
          SizedBox(
            height: 40,
            width: 40,
            child: FloatingActionButton(
              heroTag: null,
              onPressed: () {
                gridController.refresh();
              },
              child: const Icon(Icons.refresh),
            ),
          ),
        ],
      ),
    );
  }
}
